hal/halx86/smp/i386/spinup.c

```diff
@@ -0,0 +1,160 @@
+ * COPYRIGHT:   Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ *              Copyright 2021 Victor Perevertkin <victor.perevertkin@reactos.org>
+ */
+
+/* INCLUDES ******************************************************************/
+
+#include <hal.h>
+#include <smp.h>
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS *******************************************************************/
+
+extern PPROCESSOR_IDENTITY HalpProcessorIdentity;
+extern PHYSICAL_ADDRESS HalpLowStubPhysicalAddress;
+extern PVOID HalpLowStub;
+
+// The data necessary for a boot (stored inside HalpLowStub)
+extern PVOID HalpAPEntry16;
+extern PVOID HalpAPEntryData;
+extern PVOID HalpAPEntry32;
+extern PVOID HalpAPEntry16End;
+extern HALP_APIC_INFO_TABLE HalpApicInfoTable;
+
+ULONG HalpStartedProcessorCount = 1;
+
+typedef struct _AP_ENTRY_DATA
+{
+    ULONG Jump32Offset;
+    ULONG Jump32Segment;
+    PVOID SelfPtr;
+    PFN_NUMBER PageTableRoot;
+    PKPROCESSOR_STATE ProcessorState;
+    KDESCRIPTOR Gdtr;
+} AP_ENTRY_DATA, *PAP_ENTRY_DATA;
+
+typedef struct _AP_SETUP_STACK
+{
+    PVOID ReturnAddr;
+    PVOID KxLoaderBlock;
+} AP_SETUP_STACK, *PAP_SETUP_STACK;
+
+/* FUNCTIONS *****************************************************************/
+
+static
+VOID
+HalpMapAddressFlat(
+    _Inout_ PMMPDE PageDirectory,
+    _In_ PVOID VirtAddress,
+    _In_ PVOID TargetVirtAddress)
+{
+    if (TargetVirtAddress == NULL)
+        TargetVirtAddress = VirtAddress;
+
+    PMMPDE currentPde;
+
+    currentPde = &PageDirectory[MiAddressToPdeOffset(TargetVirtAddress)];
+
+    // Allocate a Page Table if there is no one for this address
+    if (currentPde->u.Long == 0)
+    {
+        PMMPTE pageTable = ExAllocatePoolZero(NonPagedPoolMustSucceed , PAGE_SIZE, TAG_HAL);
+
+        currentPde->u.Hard.PageFrameNumber = MmGetPhysicalAddress(pageTable).QuadPart >> PAGE_SHIFT;
+        currentPde->u.Hard.Valid = TRUE;
+        currentPde->u.Hard.Write = TRUE;
+    }
+    
+    // Map the Page Table so we can add our VirtAddress there (hack around I/O memory mapper for that)
+    PHYSICAL_ADDRESS b = {.QuadPart = (ULONG_PTR)currentPde->u.Hard.PageFrameNumber << PAGE_SHIFT};
+
+    PMMPTE pageTable = MmMapIoSpace(b, PAGE_SIZE, MmCached);
+
+    PMMPTE currentPte = &pageTable[MiAddressToPteOffset(TargetVirtAddress)];
+    currentPte->u.Hard.PageFrameNumber = MmGetPhysicalAddress(VirtAddress).QuadPart >> PAGE_SHIFT;
+    currentPte->u.Hard.Valid = TRUE;
+    currentPte->u.Hard.Write = TRUE;
+
+    MmUnmapIoSpace(pageTable, PAGE_SIZE);
+
+    DPRINT("Map %p -> %p, PDE %u PTE %u\n",
+           TargetVirtAddress,
+           MmGetPhysicalAddress(VirtAddress).LowPart,
+           MiAddressToPdeOffset(TargetVirtAddress),
+           MiAddressToPteOffset(TargetVirtAddress));
+}
+
+static
+PHYSICAL_ADDRESS
+HalpInitTempPageTable(
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    PMMPDE pageDirectory = ExAllocatePoolZero(NonPagedPool, PAGE_SIZE, TAG_HAL);
```

Comment:
Can't we just reuse the CR3 from the Idle process?

Reply:
I suppose, if we don’t mind the IdleProcess having a mapping of HalLowStub.

I was under the impression that wasn’t something we wanted originally.

Reply:
Since it's a temporary mapping, I don't really see any problem with it. And it can be anywhere and doesn't need to be a 1:1 mapping, right? So you can just use MmMapIoSpace or something.